src/cdk-experimental/testing/component-harness.ts

```diff
@@ -252,3 +258,75 @@ export interface ComponentHarnessConstructor<T extends ComponentHarness> {
    */
   hostSelector: string;
 }
+
+/** A class used to represent a ComponentHarness class with additional predicates applied. */
+export class HarnessPredicate<T extends ComponentHarness> {
+  private _predicates: AsyncPredicate<T>[] = [];
+  private _descriptions: string[] = [];
+
+  constructor(public harnessType: ComponentHarnessConstructor<T>) {}
+
+  /**
+   * Checks if a string matches the given pattern.
+   * @param s The string to check.
+   * @param pattern The pattern the string is expected to match. If `pattern` is a string, `s` is
+   *   expected to match exactly. If `pattern` is a regex, a partial match is allowed.
+   */
+  static stringMatches(s: string, pattern: string | RegExp): boolean {
+    return typeof pattern === 'string' ? s === pattern : !!s.match(pattern);
+  }
+
+  /**
+   * Adds a predicate function to be run against candidate harnesses.
+   * @param description A description of this predicate that may be used in error messages.
+   * @param predicate An async predicate function.
+   * @return this (for method chaining).
+   */
+  add(description: string, predicate: AsyncPredicate<T>) {
+    this._descriptions.push(description);
```

Comment:
Do we need to make sure that there are no duplicates here? Maybe we should make these arrays into sets?

Reply:
I'm not sure why there would be duplicates, I think it should be fine to leave as a list